src/btree/bt_discard.c

```diff
@@ -240,6 +240,28 @@ __free_page_modify(WT_SESSION_IMPL *session, WT_PAGE *page)
     __wt_free(session, page->modify);
 }
 
+/*
+ * __ref_addr_safe_free --
+ *     Free the ref address if we can be sure no thread is accessing it, or schedule it to be freed
+ *     otherwise.
+ */
+static void
+__ref_addr_safe_free(WT_SESSION_IMPL *session, void *ref_addr)
+{
+    WT_DECL_RET;
+    uint64_t split_gen;
+
+    split_gen = __wt_gen(session, WT_GEN_SPLIT);
```

Comment:
Ah, I see. Thanks.

Do you think this also needs the WT_FULL_BARRIER? If it does, we might need two different version of this function: one that is called in the tight loop freeing a bunch of refs (which only need a single split generation), and one which is a one-off free that doesn't want to fuss with doing the barrier/generation dance.

Reply:
I'm also happy to be convinced that a barrier isn't necessary, I didn't think that through to the ground.

Reply:
In the split logic, we save the `split_gen` in the parent page. Due to this reason, we do not want the code to be reordered, which means replacing the old page index with a new one before the `split_gen` save operation. In this ref address-free scenario, we already have the logic of replacing the `ref->addr` with NULL using an atomic operation. I do not think that we need a full barrier here.

Reply:
To avoid incrementing the `split_gen` for every ref address free, how about moving the `gen_next` into the caller and incrementing it only once? Also, it is worth mentioning that the address is freed only when it is `off page`.

Reply:
Ack. The call to __wt_gen is pretty lightweight as is It just retrieves a shared variable wihout any coordination.

I think the code is fine as-is since if it doesn't require additional coordination.